Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +20,55 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return totalAmpHoursUsed_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if (current_ < 0)
+    {
+       return true;
+    }
+    else
+    {
+       return false;
+    }
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
-}
+    int msRemaining;
+    if (isCharging())
+    {
+      msRemaining = totalAmpHoursUsed() / averageCurrent_ * HOURS_TO_MS_CONVERSION;
+    }
+    else
+    {
+      msRemaining = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed()) / averageCurrent_ * HOURS_TO_MS_CONVERSION;
+    }
+
+    return QTime(0, 0, 0, 0).addMSecs(msRemaining);
 
+}
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
+    if (previousTime_.isNull())
+    {
+        previousTime_ = batteryData.time;
+    }
+    QTime presentTime = batteryData.time;
+    int deltaMseconds = abs(presentTime.msecsTo(previousTime_));
+    double deltaHours = deltaMseconds * MS_TO_HOURS_CONVERSION;
```

Comment:
You can run into type issues here. 

See the example below.
```
double = int * double
double = int
double = int // casted as double
```
With actual numbers this would look like:
```
x = 2 * 2.2
x = 4 // answer is 4.4 but gets casted to int
x = 4.0 // The int gets converted to a double
```

The solution would be to cast the int to double before multiplying.
```
double deltaHours = (double)deltaMseconds * MS_TO_HOURS_CONVERSION;
```